### PR TITLE
Class FDR (Please do not merge, this yet)

### LINKED ISSUE #55: pyVCF reader deprecated only supported for py3.6
The python dependencies pyVCF is only supported for python <= 3.6 which make difficult to plug higher versions 3.8.

### LINKED ISSUE #56: Implementation of Class FDR
Class FDR is used to filter peptide identifications / PSMs using the target/decoy approach of the class they they represent, e.g `pseudo_X`. The method should allow filtering by specific peptide class like `pseudo_` or `COSMIC_` or by group of classes like mutations which group a set of classes [`COSMIC`, `cbio`]. 

The class-FDR should be implemented for the following file formats: 

- [x] idXML
- [x] Triqler 
- [ ] MSstats: **Note** This file format do not contain the search score for the peptide which make impossible to perform the FDR calculation.

### PR DESCRIPTION
**Pull request by ypriverol**
This PR contains the FDR-class logic to validate novel peptides: 

- [x] FDR-class implementation for idXML 
- [x] Refactoring the command line config options. 
- [x] Code cleaning.  
- [x] 0.0.19 -> 0.0.20

### CODE REVIEW

File: pypgatk/commands/dnaseq_to_proteindb.py [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
 
+log = logging.getLogger(__name__)
 
 @click.command("dnaseq-to-proteindb", short_help="Generate peptides based on DNA sequences")
-@click.option('-c', '--config_file', help='Configuration to perform conversion between ENSEMBL Files',
-              default=this_dir + '/../config/ensembl_config.yaml')
+@click.option('-c', '--config_file', help='Configuration to perform conversion between ENSEMBL Files')
 @click.option('--input_fasta', help='Path to sequences fasta')
 @click.option('--translation_table', default=1, type=int, help='Translation Table (default 1)')
 @click.option('--num_orfs', default=3, type=int, help='Number of ORFs (default 0)')
```
Comment (timosachsenberg) [inline pypgatk/commands/dnaseq_to_proteindb.py]: here it says default 0 in the comment but default=3

File: pypgatk/commands/peptide_class_fdr.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+@click.command('peptide-class-fdr', short_help="Command to compute the Peptide class FDR")
+@click.option('-c', '--config_file', help='Configuration to perform Peptide Class FDR')
+@click.option('-in', '--input_idxml', help='idxml from openms with peptides and proteins')
+@click.option('-out', '--output_idxml', help='idxml from openms with filtered peptides and proteins')
+@click.option('--min-peptide-length', help='minimum peptide length')
+@click.option('--psm-pep-fdr-cutoff', help="PSM peptide FDR cutoff or threshold", default=0.01)
+@click.option('--psm-pep-class-fdr-cutoff', help="PSM class peptide FDR cutoff or threshold", default=0.01)
+@click.option('--peptide_groups_prefix', help="Peptide class "
+              "groups e.g. \"{non_canonical:[altorf,pseudo,ncRNA];mutations:[COSMIC,cbiomut];variants:[var_mut,var_rs]}\"")
+@click.option('--peptide_classes_prefix', help='Peptides classes e.g. \"altorf,pseudo,ncRNA,COSMIC,cbiomut,var_mut,var_rs\"')
```
Comment (timosachsenberg) [inline pypgatk/commands/peptide_class_fdr.py]: maybe document how class and group prefix relate to each other e.g., do all class prefices need to be contained in the groups prefices?

File: pypgatk/commands/peptide_class_fdr.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+@click.option("--enable_class_fdr", help="Enable Class-FDR over Global PSM FDR (default true)", default = True)
```
Comment (timosachsenberg) [inline pypgatk/commands/peptide_class_fdr.py]: document how this related to the fdr parameters. e.g., is the other one ignored? if so why do we have two parameters (one for pep and one for class?

File: pypgatk/proteomics/openms.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+    :param accessions:  all protein accessions associated with the peptide.
+    :return: True if all protein accessions belows to one of these peptide_group_members.
+    """
+
+    accession_group = 0
+    for accession in accessions:
+      for class_peptide in peptide_group_members:
+        if class_peptide in accession:
+          accession_group += 1
+    return len(accessions) == accession_group
```
Comment (timosachsenberg) [inline pypgatk/proteomics/openms.py]: I hope there is no accession that matches to more than one class :)

File: pypgatk/proteomics/openms.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+    self._peptide_class_fdr_enable = self.get_default_parameters()[self.CONFIG_KEY_OPENMS_ANALYSIS][
+      self.CONFIG_PEPTIDE_APPLY_CLASS_FDR]
+    if self.CONFIG_PEPTIDE_APPLY_CLASS_FDR in self.get_pipeline_parameters():
+      self._peptide_class_fdr_enable = self.get_pipeline_parameters()[self.CONFIG_PEPTIDE_APPLY_CLASS_FDR]
+
+  @staticmethod
+  def is_peptide_group(peptide_group_members, accessions):
+    """
+    Get the True if the accessions of the peptide all belongs to the specific category.
```
Comment (timosachsenberg) [inline pypgatk/proteomics/openms.py]:
```suggestion
    Given a group of classes and a list of accessions of a peptide. Returns True if all accessions match to exactly one class in the group.
```

File: pypgatk/proteomics/openms.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    for peptide_tuple in peptide_hit_dict:
+      (psm, peptide_identification) = peptide_tuple
+      accessions = [ev.getProteinAccession() for ev in psm.getPeptideEvidences()]
+
+      if any(self._decoy_prefix in s for s in accessions):
+        decoy_count += 1
+      else:
+        target_count += 1
+
+      FDR = float(decoy_count) / float(target_count)
```
Comment (timosachsenberg) [inline pypgatk/proteomics/openms.py]: hmm wouldn't you want to calculate the q-value instead? you want to calculate the minimum FDR at that point not the FDR.

Comment (timosachsenberg) [reply pypgatk/proteomics/openms.py]: wouldn't it be more something like
```
# add columns containing FDR and the q-value respectively
def calcQ(df, scoreColName, labelColName = 'Label', classColumnName = 'class', classes=["variants", "..."]ascending = False):
    df.sort_values(scoreColName, ascending=ascending, inplace = True)   
    # global FDR
    df['FDR'] = (range(1, len(df) + 1)/df['Label'].cumsum()) - 1
    df['q-val'] = df['FDR'][::-1].cummin()[::-1]
    
    # add q-values calculated from the different classes
    ls = []
    for c in classes:
        # split the dataframe and save the subset
        currClass = pd.DataFrame(df[df[classColName] == c])
        ls.append(currClass)
            
        # calculate class-specific q-value
        currClass.sort_values(scoreColName, ascending=ascending, inplace = True)
        FDR = (range(1, len(currClass[labelColName]) + 1)/currClass[labelColName].cumsum()) - 1
        currClass['class-specific_q-val'] = FDR[::-1].cummin()[::-1]            
    df = pd.concat(ls)
    df.sort_values(scoreColName, ascending=ascending, inplace = True)
    return df
```

Comment (timosachsenberg) [reply pypgatk/proteomics/openms.py]: then I store the df back in the idXML similar to this:
```
def storeIdXML(input_file, df, output_file):
    map_id2qvalue = dict()
    for index, row in df.iterrows():
        key = row["SpecId"] + "_" + str(row["PSMId"])
        map_id2qvalue[key] = row["class-specific_q-val"]

    # load old data again
    from pyopenms import IdXMLFile, PeptideHit
    prot_ids = []; pep_ids = []
    IdXMLFile().load(input_file, prot_ids, pep_ids)

    new_pep_ids = []
    for peptide_id in pep_ids:
        hits = peptide_id.getHits()
        psmid = 1
        specid = peptide_id.getMetaValue("spectrum_reference")
        new_hits = []
        for h in hits:
            key = specid + "_" + str(psmid)
            if key in map_id2qvalue:
                h.setScore(map_id2qvalue[key]) # set class-specific q-value as main score (will be used for sorting)
                new_hits.append(h)
            psmid += 1
        peptide_id.setHits(new_hits)
        peptide_id.setScoreType("q-value")
        new_pep_ids.append(peptide_id)

    IdXMLFile().store(output_file, prot_ids, pep_ids)
```

File: pypgatk/commands/peptide_class_fdr.py (hunk 2 above)
Comment (ypriverol) [reply pypgatk/commands/peptide_class_fdr.py]: @timosachsenberg How do you think this should be documented.

File: pypgatk/proteomics/openms.py (hunk 6 above)
Comment (ypriverol) [reply pypgatk/proteomics/openms.py]: @timosachsenberg For this implementation, you need to read twice the IdXML. Do you think that will be faster than only one go. I'm wondering.

Comment (timosachsenberg) [reply pypgatk/proteomics/openms.py]: I think I did it for memory reasons (the code was part of some machine learning script). You can of course keep the unmodified peptide ids and protein ids.

Comment (ypriverol) [reply pypgatk/proteomics/openms.py]: Implementation done. Take a look @timosachsenberg

File: pypgatk/proteomics/openms.py [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+    """
+    Convert string in dataframe to Flot values
+    :param df: DataFrame
+    :return: Dataframe with all values in Flot
+    """
+    for col in df:
+      try:
+        if all([i.is_integer() for i in df[col]]):
+          df[col] = [int(x) for x in df[col]]
+      except:
```
[the remark this answers is not in the document]
Comment (ypriverol) [reply pypgatk/proteomics/openms.py]: help

File: pypgatk/proteomics/openms.py [hunk 8]
```diff
@@ -14,9 +17,29 @@ class OpenmsDataService(ParameterConfiguration):
...
   CONFIG_PEPTIDE_GROUP_PREFIX = "peptide_groups_prefix"
-  CONFIG_PEPTIDE_APPLY_CLASS_FDR = "enable_class_fdr"
+  CONFIG_PEPTIDE_DISABLE_CLASS_FDR = "disable_class_fdr"
+  CONFIG_PEPTIDE_DISABLE_BAYESIAN_FDR = "disable_bayesian_class_fdr"
 
   def __init__(self, config_file, pipeline_arguments):
+
+    self._psm_df_index = "openms_psm_index"
+    self._psm_spectrum_reference = "spectrum_reference"
+    self._openms_exclude_columns = ['SpecId', 'Label', 'ScanNr', 'Peptide',
```
Comment (timosachsenberg) [inline pypgatk/proteomics/openms.py:30]: this was very specific to some use case. here the columns you don't want to normalize or convert to numeric values need to go

Comment (timosachsenberg) [reply pypgatk/proteomics/openms.py:30]: or just leave them empty for now

File: pypgatk/proteomics/openms.py [hunk 9]
```diff
@@ -306,23 +307,38 @@ def _filter_write_idxml_with_df(self, df: DataFrame, new_columns: list, input_fi
 
     new_pep_ids = []
 
+    protein_dic = {}
+    for pro in prot_ids:
+      pro_acc = pro.getIdentifier()
```
Comment (timosachsenberg) [inline pypgatk/proteomics/openms.py]: maybe rename to pro_run_id
accession sounds like protein accession

File: pypgatk/proteomics/openms.py [hunk 10]
```diff
@@ -306,23 +307,38 @@ def _filter_write_idxml_with_df(self, df: DataFrame, new_columns: list, input_fi
...
+      ms_run = []
+      pro.getPrimaryMSRunPath(ms_run)
+      ms_run = [n.decode() for n in ms_run]
+      protein_dic[pro_acc] = "_".join(ms_run)
```
Comment (timosachsenberg) [inline pypgatk/proteomics/openms.py]: here as well

File: pypgatk/proteomics/openms.py [hunk 11]
```diff
@@ -306,23 +307,38 @@ def _filter_write_idxml_with_df(self, df: DataFrame, new_columns: list, input_fi
...
+
     for peptide_id in pep_ids:
       hits = peptide_id.getHits()
       psmid = 1
       specid = peptide_id.getMetaValue(self._psm_spectrum_reference)
+
+      pep_acc = peptide_id.getIdentifier()
```
Comment (timosachsenberg) [inline pypgatk/proteomics/openms.py:343]: same here

File: pypgatk/proteomics/openms.py [hunk 12]
```diff
@@ -306,23 +307,38 @@ def _filter_write_idxml_with_df(self, df: DataFrame, new_columns: list, input_fi
...
+      if pep_acc not in protein_dic:
+        raise ValueError(
+          "The reference peptide in the PeptideIdentification {}--{} can't be found in the ProteinIdentifications".format(
+            pep_acc, specid))
+
+      ms_run_acc = protein_dic[pep_acc]
```
Comment (timosachsenberg) [inline pypgatk/proteomics/openms.py:349]: and here

File: pypgatk/proteomics/openms.py [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+
+    df_psms.sort_values("score", ascending=ascending, inplace=True)
+    df_psms['FDR'] = (range(1, len(df_psms) + 1) / df_psms['target'].cumsum()) - 1
+    df_psms['q-value'] = df_psms['FDR'][::-1].cummin()[::-1]
+
+    df_psms.sort_values("score", ascending=ascending, inplace=True)
+
+    return df_psms
+
+  def _compute_bayesian_class_fdr(self, df_psms: DataFrame):
```
Comment (timosachsenberg) [inline pypgatk/proteomics/openms.py:149]: this is not done yet right?

**Review by timosachsenberg — COMMENTED** [review]
looks good

File: pypgatk/proteomics/openms.py (hunk 13 above)
Comment (ypriverol) [reply pypgatk/proteomics/openms.py:149]: Yes is done ;)

**Comment by ypriverol** [comment]
- Number of PSM after Global FDR filtering: **7563**
- Number of PSM after Non-bayesian FDR filtering: **7434**
- Number of PSM after Bayesian FDR filtering: **7563**

**Comment by timosachsenberg** [comment]
Exact same value for Bayesian? Ist that by design

**Comment by ypriverol** [comment]
To be honest, it shouldn't be, but I'm not expert in the Bayesian method. I will wait for the review of @yafeng @husensofteng 
 for feedback.

File: pypgatk/commands/peptide_class_fdr.py [hunk 14]
```diff
@@ -17,7 +17,7 @@
 @click.option('-in', '--input-file', help='input file with the peptides and proteins')
 @click.option('-out', '--output-file', help='idxml from openms with filtered peptides and proteins')
 @click.option("--file-type", default = 'idxml')
-@click.option('--min-peptide-length', help='minimum peptide length')
+@click.option('--min-peptidopenmspeakpickere-length', help='minimum peptide length')
```
Comment (timosachsenberg) [inline pypgatk/commands/peptide_class_fdr.py]: error?

**Pull request merged**